#### This assignment is taken from the repo "https://github.com/prof-sushil/Python-zero2hero"

# Assignment 2.1
* Develop a calculator to take any two numbers from user and perform the operations on those two numbers (operations to be performed are chosen by user). 

* Add as much functionality as you could that we have learnt till now.

In [ ]:
# Simple calculator

num1 = int(input("Enter first number: "))
num2 = int(input("Enter second number: "))

print("1. Add")     
print("2. Subtract")
print("3. Multiply")
print("4. Divide")
print("5. Exponent")
print("6. Remainder")
print("7. floor division")

choice = int(input("Enter choice: "))
if choice == 1:
    print(f"{num1} + {num2} = ", num1 + num2)  #addition
elif choice == 2:
    print(f"{num1} - {num2} = ", num1 - num2)  #subtraction
elif choice == 3:
    print(f"{num1} * {num2} = ", num1 * num2)  #multiplication
elif choice == 4:
    print(f"{num1} / {num2} = ", num1 / num2)  #division
elif choice == 5:
    print(f"{num1} ^ {num2} = ", num1 ** num2)  #exponent
elif choice == 6:
    print(f"{num1} % {num2} = ", num1 % num2) #remainder
elif choice == 7:
    print(f"{num1} // {num2} = ", num1 // num2) #floor division
else:
    print("Invalid choice")
    
    


    


# Assignment 2.2

1. Given marks of students of a class: marks = [ 10, 20, 30, 40, 50, 60]. Calculate the average mark.

2. Write a program to pring multiplication table of a number given by user.

3. Given an integer x, return true if x is a palindrome, and false otherwise.

4. Given a list of numbers, find the greatest number.

5. Given a list of integers x, find the peak elements. A peak is an element that is greater than its neighbors.

        Example: x = [1, 2, 3, 1, 10, 8]
        Output: peaks = [3, 10]

6. Write a program to count the number of vowels in a string.

7. Write a program to find the sum of digits in an integer.

8. Given a roman numeral, convert it to an integer. Roman numerals are represented by seven different symbols: I, V, X, L, C, D and M.

        Input: s = "LVIII"
        Output: 58
        Explanation: L = 50, V= 5, III = 3

9. Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid. An input string is valid if:

* Open brackets must be closed by the same type of brackets.
* Open brackets must be closed in the correct order.
* Every close bracket has a corresponding open bracket of the same type.

        Example 1:
        Input: s = "()"
        Output: true

        Example 2:
        Input: s = "()[]{}"
        Output: true

        Example 3:
        Input: s = "(]"
        Output: false

10. Write a program to calculate the frequency of each characters in a string.

        Example:
        Input: s = "hello"
        Output: {
                'h': 1,
                'e': 1,
                'l': 2,
                'o': 1
        }

11. Write a program to generate percentage of each students whose records are given below:

        grades_dict = {
                "Alice": {"Math": 90, "Science": 85, "Literature": 88},
                "Bob": {"Math": 78, "Science": 82, "Literature": 80},
                "Charlie": {"Math": 92, "Science": 91, "Literature": 94}
        }

        Output: 
                Alice: 87.67
                Bob: 80.00
                Charlie: 92.33




In [ ]:
# 1. print average marks
marks = [10,20,30,40,50,60]
sum = 0.0
for mark in marks:
    sum = sum + mark
    
avg_mark = sum / len(marks)
print(f"Average of marks is {avg_mark}")



Average of marks is 35.0


In [3]:
# 2.print multiplication tabel

num = int(input("Enter the number that you want to print the table of: "))
for i in range(1,11):
    print(f"{num} * {i} = {num*i}")

4 * 1 = 4
4 * 2 = 8
4 * 3 = 12
4 * 4 = 16
4 * 5 = 20
4 * 6 = 24
4 * 7 = 28
4 * 8 = 32
4 * 9 = 36
4 * 10 = 40


In [5]:
#3. check  palindrome for a integer x and return true otherwise false

x = int(input(" Enter the number that you want to check for palindrome "))
print("Given number is: ", x)
x_str = str(x)
rev_str = x_str[::-1]
if rev_str == x_str:
    result =  True
else:
    result = False
print(f"Palindrome = {result}")

Given number is:  1221
Palindrome = True


In [6]:
#4. given a list of number find greatest number
list = [2,4,5,3,6,7,9,45,22,-1]
key = list[0]
for i in list:
    if key < i:
        key = i
print(f"The greatest number in the list is {key}")

The greatest number in the list is 45


In [8]:
#5. 
x = [1,2,3,1,10,8]
item = x[0]
peak = []
for i in range(1,len(x)-1):
    # print(i)
    if x[i] > x[i - 1] and x[i] > x[i + 1]:
        peak.append(x[i])
print(peak)


[3, 10]


In [9]:
#6. 
vowel = ['a','e','i','o','u','A','E',"I","O",'U']
str = 'aeiouAbc'
count = 0
for char in str:
    if char in vowel:
        count += 1
print(f"Number of vowels in the string '{str}' is {count}")

Number of vowels in the string 'aeiouAbc' is 6


In [31]:
#7. 
# del str

nums = int(input("Enter a number"))
print(f"Given number is: {nums}")
str_num = str(nums)
sums = 0
length = len(str_num)

for i in range(0,length):
    mod = nums % 10
    sums = sums + mod
    nums = nums // 10
print(f"The sum of digits is {sums}")

Given number is: 123
The sum of digits is 6


In [ ]:
#8. 
# Roman = {'I' : 1,
#          'V' : 5,
#          'X' : 10,
#          'L' : 50,
#          'C' : 100,
#          'D': 500,
#          'M': 1000
#          }
# input_str = input(" Enter a string in roman numeral to convert to integer ")
# input_str = "LVIII"
# sum = 0


In [17]:
#9. 
def is_valid_bracket_sequence(s):
    # Dictionary to map opening and closing brackets
    bracket_map = {')': '(', '}': '{', ']': '['}
    stack = []

    for char in s:
        if char in bracket_map.values():  # If it's an opening bracket
            stack.append(char)
        elif char in bracket_map:  # If it's a closing bracket
            if not stack or stack[-1] != bracket_map[char]:  # Stack is empty or mismatch
                return "The given string is not valid"
            stack.pop()  # Valid match, remove the last opening bracket
        else:
            return "The given string is not valid"  # Invalid character

    return "The given string is valid" if not stack else "The given string is not valid"
 # Test cases
s = "()[]{}"
print(is_valid_bracket_sequence(s))  # Output: The given string is valid

s = "(){]"
print(is_valid_bracket_sequence(s))  # Output: The given string is not valid


The given string is valid
The given string is not valid


In [24]:
#10. 
s = "hello"
def count_frequency(s):
    frequency = {}
    for char in s:
        if char in frequency:
            frequency[char] += 1
        else:
            frequency[char] = 1
    return frequency
print(count_frequency(s))  




{'h': 1, 'e': 1, 'l': 2, 'o': 1}


In [ ]:
#11.
del sum

grades_dict = {
    "Alice": {"Math": 90, "Science": 85, "Literature": 88},
    "Bob": {"Math": 78, "Science": 82, "Literature": 80},
    "Charlie": {"Math": 92, "Science": 91, "Literature": 94}
}

for student, subjects in grades_dict.items():
    total = sum(subjects.values())
    percentage = (total / (len(subjects) * 100)) * 100  # Assuming each subject is out of 100
    print(f"{student}: {percentage:.2f}")


Alice: 87.67
Bob: 80.00
Charlie: 92.33


### Assignment 3.1

1. You are given a large integer represented as an integer list of digits, where each digits[i] is the ith digit of the integer. The digits are ordered from most significant to least significant in left-to-right order. The large integer does not contain any leading 0's. Increment the large integer by one and return the resulting ;list of digits. Implement the logic by creating a function.

        Example 1:
        Input: digits = [1,2,3]
        Output: [1,2,4]
        Explanation: The array represents the integer 123.
        Incrementing by one gives 123 + 1 = 124.
        Thus, the result should be [1,2,4].

        Example 2:
        Input: digits = [4,3,2,1]
        Output: [4,3,2,2]
        Explanation: The array represents the integer 4321.
        Incrementing by one gives 4321 + 1 = 4322.
        Thus, the result should be [4,3,2,2].

        Example 3:
        Input: digits = [9]
        Output: [1,0]
        Explanation: The array represents the integer 9.
        Incrementing by one gives 9 + 1 = 10.
        Thus, the result should be [1,0].
        ]

### Assignment 3.2
1. Use a custom defined timeit decorator to measure the execution time of any function. Use time module to compute time.
        
        Hint:
        import time
        current_time = time.time()

### Assignment 4.1

1. Create a class Rectangle. Define the required object attributes/state yourself. Define class methods: area() and perimeter() and use them by creating an instance of rectangle.

2. You are given an integer list height of length n. There are n vertical lines drawn such that the two endpoints of the ith line are (i, 0) and (i, height[i]).Find two lines that together with the x-axis form a container, such that the container contains the most water.Return the maximum amount of water a container can store. Notice that you may not slant the container.

Example1:

<div>
<img src="water-container-problem.png" width="500"/>
</div>

    Input: height = [1,8,6,2,5,4,8,3,7]

    Output: 49

Explanation: The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7]. In this case, the max area of water (blue section) the container can contain is 49.


Example2:

    Input: height = [1,1]

    Output: 1




### Assignment 5.1

1. I have provided the students.json inside data_files folder that contains students records. Using that json, compute following things:

* Calculate the percentage for each students assuming 100 is the full marks for each subjects. Also create a new json file with additional field "percentage" in json whose value will be the percentage you have computed.

* Then load the new json file and generate the statistics of:
    * How many students failed? (i.e. percentage < 40)
    * How many students achieve third division (i.e. percentage between 40 and 50)
    * How many students achieve second division (i.e. percentage between 50 and 65)
    * How many students achieve first division (i.e. percentage between 65 and 80)
    * How many students achieve distinction division (i.e. percentage 80 and above )

